File: project/donghyeon-onboarding/src/test/java/me/dhlee/donghyeononboarding/service/SurveyServiceTest.java
package me.dhlee.donghyeononboarding.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import me.dhlee.donghyeononboarding.domain.ItemType;
import me.dhlee.donghyeononboarding.domain.Survey;
import me.dhlee.donghyeononboarding.dto.request.SurveyCreateRequest;
import me.dhlee.donghyeononboarding.dto.request.SurveyItemCreateRequest;
import me.dhlee.donghyeononboarding.dto.request.SurveyItemOptionCreateRequest;
import me.dhlee.donghyeononboarding.exception.AppException;
import me.dhlee.donghyeononboarding.exception.ErrorCode;
import me.dhlee.donghyeononboarding.repository.SurveyRepository;

@SpringBootTest
class SurveyServiceTest {

    @Autowired
    protected SurveyService surveyService;

    @Autowired
    protected SurveyRepository surveyRepository;

    @BeforeEach
    void setUp() {
        surveyRepository.deleteAllInBatch();
    }

    @DisplayName("설문조사를 생성할 수 있다.")
    @Test
    void createSurvey() {
        var option = new SurveyItemOptionCreateRequest("옵션1", 0);
        var item = new SurveyItemCreateRequest("질문1", "질문 설명1", ItemType.MULTI_SELECT, true, 0, List.of(option));
        var request = new SurveyCreateRequest("설문 제목", "설문 설명", List.of(item));

        var result = surveyService.createSurvey(request);

        assertThat(result).isNotNull();

        List<Survey> all = surveyRepository.findAll();
        assertThat(all).hasSize(1);
        assertThat(all.get(0).getTitle()).isEqualTo("설문 제목");
        assertThat(all.get(0).getDescription()).isEqualTo("설문 설명");
    }
}

File: project/doohyeon-onboarding/src/test/kotlin/com/example/doohyeononboarding/api/service/SurveyServiceTest.kt
package com.example.doohyeononboarding.api.service

import com.example.doohyeononboarding.api.service.request.CreateSurveyRequest
import com.example.doohyeononboarding.api.service.request.QuestionRequest
import com.example.doohyeononboarding.domain.question.QuestionType
import com.example.doohyeononboarding.domain.servey.Survey
import com.example.doohyeononboarding.domain.servey.SurveyRepository
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class SurveyServiceTest @Autowired constructor(
    private val surveyService: SurveyService,
    private val surveyRepository: SurveyRepository,
) {

    @AfterEach
    fun tearDown() {
        surveyRepository.deleteAll()
    }

    @DisplayName("설문 항목이 0개인 경우 예외가 발생한다.")
    @Test
    fun createSurvey_Fail1() {
        // given
        val request = CreateSurveyRequest(
            title = "설문조사 제목",
            description = "설문조사 설명",
            questions = emptyList()
        )

        // when // then
        assertThatThrownBy { surveyService.createSurvey(request) }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("설문 항목은 1개 이상 10개 이하이어야 합니다.")
    }

    @DisplayName("설문 항목이 11개인 경우 예외가 발생한다.")
    @Test
    fun createSurvey_Fail2() {
        // given
        val request = CreateSurveyRequest(
            title = "설문조사 제목",
            description = "설문조사 설명",
            questions = (1..11).map {
                QuestionRequest(
                    title = "질문 $it",
                    description = "설명 $it",
                    type = QuestionType.SHORT_TEXT,
                    isRequired = true,
                    options = null
                )
            }
        )

        // when // then
        assertThatThrownBy { surveyService.createSurvey(request) }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("설문 항목은 1개 이상 10개 이하이어야 합니다.")
    }

    @DisplayName("설문 조사를 정상적으로 생성된다.")
    @Test
    fun createSurvey_Success() {
        // given
        val request = CreateSurveyRequest(
            title = "설문조사 제목",
            description = "설문조사 설명",
            questions = listOf(
                QuestionRequest(
                    title = "이름",
                    description = "이름을 입력하세요",
                    type = QuestionType.SHORT_TEXT,
                    isRequired = true,
                    options = null
                )
            )
        )
        surveyRepository.save(
            Survey(
                title = request.title,
                description = request.description,
            )
        )

        // when
        val response = surveyService.createSurvey(request)

        // then
        assertThat(response.surveyId).isNotNull()
    }

}
